types/invoker.go

```diff
@@ -15,55 +15,77 @@ type Invoker struct {
 	GatewayURL    string
 }
 
-func (i *Invoker) Invoke(topicMap *TopicMap, topic string, message *[]byte) {
-	if len(*message) > 0 {
+type InvocationResult struct {
+	StatusCode int
+	Body       *[]byte
+	Error      error
+}
 
-		matchedFunctions := topicMap.Match(topic)
-		for _, matchedFunction := range matchedFunctions {
+func NewInvocationResult(statusCode int, body *[]byte, error error) *InvocationResult {
+	return &InvocationResult{StatusCode: statusCode, Body: body, Error: error}
+}
 
-			log.Printf("Invoke function: %s", matchedFunction)
+type InvocationResponse struct {
+	StatusCode int
+	Body       *[]byte
+	Headers http.Header
+}
 
-			gwURL := fmt.Sprintf("%s/function/%s", i.GatewayURL, matchedFunction)
-			reader := bytes.NewReader(*message)
+func NewInvocationResponse(statusCode int, body *[]byte, headers http.Header) *InvocationResponse {
+	return &InvocationResponse{StatusCode: statusCode, Body: body, Headers: headers}
+}
 
-			body, statusCode, doErr := invokefunction(i.Client, gwURL, reader)
+func (i *Invoker) Invoke(topicMap *TopicMap, topic string, message *[]byte) (result map[string]*InvocationResult) {
+	result = make(map[string]*InvocationResult)
 
-			if doErr != nil {
-				log.Printf("Unable to invoke from %s, error: %s\n", matchedFunction, doErr)
-				return
-			}
+	if message != nil && len(*message) > 0 {
 
-			printBody := false
-			stringOutput := ""
+		matchedFunctions := topicMap.Match(topic)
+		for _, matchedFunction := range matchedFunctions {
 
-			if body != nil && i.PrintResponse {
-				stringOutput = string(*body)
-				printBody = true
-			}
+			log.Printf("Invoke function: %s", matchedFunction)
+			functionURL := fmt.Sprintf("%s/function/%s", i.GatewayURL, matchedFunction)
+			reader := bytes.NewReader(*message)
 
-			if printBody {
-				log.Printf("Response [%d] from %s %s", statusCode, matchedFunction, stringOutput)
+			response, err := i.performInvocation(functionURL, reader)
 
+			if err != nil {
+				if response != nil {
+					result[matchedFunction] = NewInvocationResult(response.StatusCode, nil, err)
+				} else {
+					result[matchedFunction] = NewInvocationResult(-1, nil, err)
+				}
 			} else {
-				log.Printf("Response [%d] from %s", statusCode, matchedFunction)
+				result[matchedFunction] = NewInvocationResult(response.StatusCode, response.Body, err)
+			}
+
+			if response != nil && response.Body != nil && i.PrintResponse {
+				stringOutput := string(*response.Body)
+				log.Printf("Headers: %s", response.Headers)
+				log.Printf("Response: [%d] from %s %s", response.StatusCode, matchedFunction, stringOutput)
 			}
 		}
 	}
+	return result
 }
 
-func invokefunction(c *http.Client, gwURL string, reader io.Reader) (*[]byte, int, error) {
+func (i *Invoker) performInvocation(functionURL string, bodyReader io.Reader) (*InvocationResponse, error) {
```

Comment:
Please revert this to the more natural sounding `invokefunction`, perhaps capitalize to `invokeFunction` or just `invoke`.